File: packages/rt_rust/core/core/source/recognizer/stack.rs
const stack_size: usize = 64;

// ///////////////////////////////////////////
// KERNEL STATE
// ///////////////////////////////////////////

pub struct KernelStack {
    pub state_stack: [u64; stack_size],
    pub stack_pointer: i32,
}

impl KernelStack {
    pub fn new() -> KernelStack {
        KernelStack {
            state_stack: [0; stack_size],
            stack_pointer: 0,
        }
    }
    pub fn reset(&mut self, kernel_state: u32) {
        self.stack_pointer = 0;

        self.state_stack[0] = 0;

        self.push_state(kernel_state)
    }

    pub fn push_state(&mut self, kernel_state: u32) {
        self.stack_pointer += 1;
        let sp = self.stack_pointer as usize;

        self.state_stack[sp] = (kernel_state as u64);

        // self.meta_stack[sp] = (self.meta_stack[sp - 1] & 0xFFFF) | (0 as u32);
    }

    pub fn swap_state(&mut self, kernel_state: u32) {
        self.state_stack[self.stack_pointer as usize] = kernel_state as u64;
    }

    pub fn pop_state(&mut self) -> u32 {
        let state = self.read_state();
        self.stack_pointer -= 1;
        return state;
    }

    pub fn read_state(&self) -> u32 {
        (self.state_stack[self.stack_pointer as usize] & 0xFFFF_FFFF) as u32
    }

    pub fn copy_state_stack(&self, dest: &mut KernelStack) {
        for i in 0..=self.stack_pointer {
            dest.state_stack[i as usize] = self.state_stack[i as usize];
        }
    }
}

File: packages/rt_typescript/source/recognizer/stack.ts
/* 
 * Copyright (C) 2021 Anthony Weathersby - The Hydrocarbon Parser Compiler
 * see /source/typescript/hydrocarbon.ts for full copyright and warranty 
 * disclaimer notice.
 */

/////////////////////////////////////////////
// Kernel State 
/////////////////////////////////////////////


export class KernelStack {
    pointer: number;
    state_stack: Uint32Array;
    meta_stack: Uint32Array;

    constructor() {

        this.state_stack = new Uint32Array(128);

        this.meta_stack = new Uint32Array(128);

        this.pointer = 0;
    }

    reset(state: number) {
        this.pointer = 0;
        this.state_stack[0] = 0;
        this.meta_stack[0] = 0;
        this.push_state(state, 0);
    }

    push_state(kernel_state: number, accumulated_symbols: number) {
        this.state_stack[++this.pointer] = kernel_state;
        this.meta_stack[this.pointer] = /* (this.meta_stack[this.pointer - 1] & 0xFFFF) | */ accumulated_symbols;
    }

    swap_state(kernel_state: number) {
        this.state_stack[this.pointer] = kernel_state;
    }

    pop_state(): number {
        let state = this.read_state();
        this.pointer--;
        return state;
    }

    read_state(): number {
        return this.state_stack[this.pointer];
    }

    copy_state_stack(dest: KernelStack) {
        for (let i = 0; i <= this.pointer; i++)
            dest.state_stack[i] = this.state_stack[i];

        dest.pointer = this.pointer;
    }

    clone(): KernelStack {
        const new_stack = new KernelStack;

        this.copy_state_stack(new_stack);

        return new_stack;
    }

    debug_clone(): KernelStack {
        const new_stack = new KernelStack;

        new_stack.state_stack = new Uint32Array(this.pointer + 1);
        new_stack.meta_stack = new Uint32Array(this.pointer + 1);

        for (let i = 0; i <= this.pointer; i++)
            new_stack.state_stack[i] = this.state_stack[i];

        new_stack.pointer = this.pointer;

        return new_stack;
    }
}
